src/app/store/observeMiddleware.ts

```diff
@@ -0,0 +1,76 @@
+import { deepEqual } from 'fast-equals';
+import { Dispatch, isAction, Middleware, MiddlewareAPI } from 'redux';
+import { createAction, createCustomAction, isActionOf } from 'typesafe-actions';
+import { RootState } from './types';
+
+interface StoreObserver<T> {
+  id: string;
+  /**
+   * Function to create "something" which will be used to determine if the side effect should run.
+   * deepEqual is used for the comparison, so use primitives, shallow objects or arrays to keep
+   * performance reasonable.
+   */
+  getObserved: (rootState: RootState) => T;
+  /**
+   * Runs the side effect providing both the previous and current version of the derrived state.
+   */
+  sideEffect: (states: { previous: T; current: T }) => void;
+}
+
+// Need to user a higher order function to get the correct typings and inference, it will now correctly
+// type the value of T based on what the getObserved returns
+export const observe = <T>(storeObserver: StoreObserver<T>) =>
+  createCustomAction('observer/OBSERVE', (storeObserver: StoreObserver<T>) => ({ storeObserver }))(
+    storeObserver,
+  );
+export const unobserve = createAction('observer/UNOBSERVE')<string>();
+export const clearObservers = createAction('observer/CLEAR_OBSERVERS')();
+
+function isObserveAction(action: unknown): action is ReturnType<typeof observe> {
+  return (
+    typeof action === 'object' &&
+    action !== null &&
+    'type' in action &&
+    action.type === 'observer/OBSERVE'
+  );
+}
+
+export function observerMiddleware<D extends Dispatch>(
+  api: MiddlewareAPI<D, RootState>,
+): ReturnType<Middleware> {
+  const observers = new Map<string, StoreObserver<unknown>>();
+  return (next) => (action) => {
+    if (!isAction(action)) {
+      return next(action);
+    }
+
+    if (isObserveAction(action)) {
+      observers.set(action.storeObserver.id, action.storeObserver);
+      return;
+    }
+
+    if (isActionOf(unobserve, action)) {
+      observers.delete(action.payload);
+      return;
+    }
+
+    if (isActionOf(clearObservers, action)) {
+      observers.clear();
+      return;
+    }
+
+    const previousRootState = api.getState();
+    const result = next(action);
+    const currentRootState = api.getState();
+
+    for (const [_id, observer] of observers) {
+      const previous = observer.getObserved(previousRootState);
+      const current = observer.getObserved(currentRootState);
+      if (deepEqual(previous, current)) {
```

Comment:
Ended up going with deep equal, since fast equals seems to be pretty fast. I think documentation on trying to keep `T` shallow should be enough.

Reply:
I think I'd maybe prefer defaulting to `Object.is` and maybe allowing the observer to override the equality function.

Also wouldn't you want to only run the effect when the old and new state are *not* equal?

Reply:
Yeah typo which is curiously working. 

I'll change the equality check method when I clean it up for proper review

Reply:
> Yeah typo which is curiously working.

Strict mode would see it working in dev actually.